### ISSUE TITLE
Microstuttering on pcexhumed and rednukem

### ISSUE THREAD

**Issue by Kappa971** [nukeykt/NBlood#468]
Hi, I have microstuttering with pcexhumed and rednukem (last builds and older builds), both in opengl and software mode. It's weird because nblood and eduke32 work just fine. My PC: CPU ryzen 5 1600, GPU gtx 1060 6gb, RAM 16gb, 144hz freesync monitor. No FPS cap, FPS offset 0, freesync on, vsync on (on Nvidia control panel). I also tried using the commands win_systemtimermode 0-1 and win_performancemode 0-1 but without results and vsync on and off

**Comment by Kappa971** [comment]
OS: Windows 10 20h2

**Comment by vanfanel** [comment]
I am also seeing this in both X86_64 GNU/Linux with AMD graphics, and in aarch64 GNU/Linux.
Strangely, it doesn't seem to be a CPU usage problem.

**Comment by CommonLoon102** [comment]
> Strangely, it doesn't seem to be a CPU usage problem according to TOP.
How so? The game should peak out a single core to 100% anyways, since it is using busy wait in the main game loop.

**Comment by vanfanel** [comment]
@CommonLoon102 You are right! Looking at all the cores using "1", I can see there's one of them permanently near 100%!

**Comment by Kappa971** [comment]
So what? Why do eduke32 and nblood run smoothly while rednukem and pcexhumed don't? I use the same settings for everyone

**Comment by CommonLoon102** [comment]
Probably because pcexhumed's and rednukem's main game loop is a mess. It looks like for me that they can't create new video frames while they are processing and waiting for user input. It is just a guess though. NBlood's and EDuke32's stuttering was solved when they rewrote the main game loop and made the game world updates unrelated to the video frame renderings.

**Comment by Kappa971** [comment]
Is there a chance to see these improvements in pcexhumed and rednukem in the future as well?

**Comment by CommonLoon102** [comment]
I think there is lack of interest of doing this. If there was, it has been done already.

**Comment by Kappa971** [comment]
Ok thanks :)

**Comment by CommonLoon102** [comment]
I use a 144Hz screen, without v-sync and FPS limit set to 144. I'm also on Win7, and as far as I know, this stuttering is Win10 specific, because it has something to do with SDL's newer versions "communicating" with Win10, but I might be totally wrong on this. I just don't have any stuttering.

**Comment by sirlemonhead** [comment]
I haven't seen this issue at all myself, though my machine is quite old so I'm certainly not going beyond 60hz.

**Comment by Kappa971** [comment]
> Maybe give Linux a try ;) You can flash Fedora easily onto a USB drive, boot into it, you don't even have to install it, try the games in question and if it doesn't stutter, then it is a Win10 issue.

I'll do a test with Linux, to understand if it is a Windowns 10 issue, Thanks

**Comment by Kappa971** [comment]
I downloaded Ubuntu, I created a bootable usb drive with rufus with persistent partition size 6gb, I installed the required packages, built rednukem, I tried to start rednukem with the required files (duke3d.grp and duke.rts) via terminal, a once clicked on "start" it crashes. I installed the Nvidia drivers, rebooted and at that point the usb drive no longer booted. Ubuntu and Linux distro zero usability :) (my opinion). On Windows you download the already built program or build it (with vcpkg and visual studio), double click on the exe and it works. Anyway I trust you if you say that the stuttering problem is not present on Linux, I hope that one day it will be solved on Windows as well. Thanks for your availability and your work :)

**Comment by CommonLoon102** [comment]
Just to be sure, I have to ask: have you installed it onto an SD card (with an USB SD card reader)? Because SD cards will go corrupt very soon, they cannot run OSes. If you used a normal pendrive or an SSD then it is fine and then it is a software issue.
Thanks your time that you have tried it at least. You could try Fedora as well instead of Ubuntu, maybe you would have better luck with it, sometimes I do. But only if you want to kill some more time with this :)

**Comment by Kappa971** [comment]
> Just to be sure, I have to ask: have you installed it onto an SD card (with an USB SD card reader)?

I used a normal pendrive.

> You could try Fedora as well instead of Ubuntu, maybe you would have better luck with it, sometimes I do.

I'll make another try with Fedora, thanks

**Comment by Kappa971** [comment]
So, I tried it with Fedora. Rednukem crashes on startup here as well. I tried pcexhumed, it starts up, it works but micro stuttering is also present here, as on Windows 10. I would like to point out that it is nothing that makes the game unplayable but it is present, the game is not as smooth as eduke32 or nblood. At this point I would say it also affects Linux. It's very strange

**Comment by CommonLoon102** [comment]
Thanks, good feedback. Then it is not just Win10. Not a big surprise, but then it might be hardware related as well. I might need to check rednukem from under Fedora. I'll give it a try some time.

**Comment by CommonLoon102** [comment]
So I have done the testing. Fedora 34, default nouveau drivers, GTX 970, 144Hz monitor, vsync on. No tearing, no stuttering, no crashes. I checked both duke3d and rr, but I don't have exhumed. Compiled from latest master (https://github.com/nukeykt/NBlood/commit/ab1984d8f741e6896ad20df2a1f770db05ca8cd5). I couldn't try without vsync, because the game didn't let me, when I set it to off, it switched back to `adaptive`.

**Comment by Kappa971** [comment]
Very strange... I also compiled from the last master but rednukem crashed on boot in both Ubuntu 21.04 and Fedora 34. I'm puzzled. Regarding stuttering, perhaps it is not the correct term, I don't know how to describe it, the game is simply not 100% smooth as nblood and eduke32 are (in pchexumed the problem is more marked than in rednukem). Do you developers all have Intel CPUs? Because I also have an Nvidia card (GTX 1060), maybe there is something wrong with rednukem/pcexhumed and AMD CPUs? Anyway I don't want to drive you crazy :) If you don't have these problems, obviously you can't solve them

**Comment by CommonLoon102** [comment]
Now I tried exhumed, and it is fine too. Also, now I know why I couldn't change the vsync settings, it is locked by the video driver and it doesn't let me to change it. With `r_maxfps 0` and `r_vsync 1` I don't have any stuttering, the fps is constant 144.0.

**Comment by Kappa971** [comment]
There are some interesting news. I did further tests with two other PCs. The first is another PC with an amd ryzen 1600 CPU, AMD Radeon RX 570 GPU and 75 hz freesync monitor, the second PC has an Intel Core I7 860 CPU (first generation), Nvidia GeForce GTX 660 GPU and a normal monitor at 60 hz. On both PCs rednukem starts up (So there must be some problem between my PC and Linux). As for stuttering, it is present in the AMD PC while on the Intel PC the game is very smooth (despite having only a 60 hz monitor). So the stuttering problem either concerns a problem between these two games and freesync or with AMD CPUs

**Comment by CommonLoon102** [comment]
Can you swap the monitors, so we can rule out the freesync case?

**Comment by Kappa971** [comment]
The GTX 660 doesn't support freesync

**Comment by CommonLoon102** [comment]
You can still try AMD CPU with monitor without freesync (or just turn off freesync).

**Comment by Kappa971** [comment]
True, I can try

**Comment by Kappa971** [comment]
I connected my PC to the 60hz monitor. rednukem crashes all the same but i could try pcexhumed and there is no stuttering or maybe it is inaudible (all tests were done with Fedora 34)

**Comment by CommonLoon102** [comment]
Then the issue title can be changed to (Freesync) instead of (Windows 10, Linux).

**Comment by Kappa971** [comment]
If anyone else could try these two games with a freesync monitor, we would have the confirmation

**Comment by CommonLoon102** [comment]
By the way, how raze performs for these games? Any stuttering with freesync?

**Comment by Kappa971** [comment]
Raze It has no stuttering problems

**Comment by Kappa971** [comment]
I don't know if this is the right place but I want to point out that VoidSW also has some problems, it doesn't seem "smooth" too.

**Comment by CommonLoon102** [comment]
Frame pacing is a neutral expression. There is good and bad frame pacing.

**Comment by Kappa971** [comment]
If it can be misleading, I can remove "frame-pacing" from the title. Either way, there is something wrong with the Build Engine, something that has been fixed in Raze (maybe Eduke32 too). When I had a 60hz monitor I didn't notice any problems, now with this 144hz freesync I have all these problems of stuttering or whatever. I'm not the only one on the internet with these problems (although I don't know if these people own regular monitors or freesync), some have switched to Raze

**Comment by symbios24** [comment]
Hi there i have the same issue in a Raspberry pi 4,in my Raspberry pi 3 works fine though
Very strange

**Comment by Kappa971** [comment]
This ticket has become quite confusing and I don't think it will be considered by the developers in this state. Perhaps a new problem should be opened. Making a summary I can say that:
1. Trying the latest build of Nblood, I notice that it has the exact same behavior as Eduke32, and save the configuration again (maybe the build I tried a month ago had some problems). Strangely I noticed that locking the fps at 120 seems smoother than 144 fps but it is my feeling and should not be taken into consideration so as not to further complicate things.
2. It has been tested that with 60hz monitors there are no problems.
3. With monitors with a refresh rate higher than 60hz with freesync (but I am convinced that it also happens in monitors without freesync, this is a test someone should do), PCExhumed and Rednukem have microstuttering problems, nothing that makes the games unplayable, walking sideways you notice that the front image snaps slightly. Here it is said that some changes have probably been made to Eduke32, which have probably been ported to Nblood but not to PCExhumed and Rednukem.
4. VoidSW I think is still in beta, last time it didn't even support vsync. Here what I notice is not stuttering but it seems as if the image is flickering (I don't know how to describe it).

The tests were done on both Windows and Linux, with LG Freesync 144hz monitors
